Week2/todo_app/program.js

```diff
@@ -23,36 +37,141 @@ class TodoModel {
 
     create(description, callback) {
         // Write code and query to create a new TODO item
+        const addTodoItem = "INSERT INTO todo_items VALUES(NULL, ?, ?, ?)";
+        this.dbConnection.query(addTodoItem, description, (err, results, fields) => {
+            if(err) return callback(err);
+            callback(null, results);
+        });
     }
 
     update(id, description, callback) {
         // Write code and query to update and existing TODO item
+        const updatePromise = this.availableIds(this.sqlSelectQuery);
+        updatePromise
+            .then(data => {
+                const temp = data.includes(id);
+                if(temp) {
+                    const updateItem = `UPDATE todo_items SET text = ? WHERE id = ?`;
+                    this.dbConnection.query(updateItem, [description, id], (err, results, fields) => {
+                    if(err) return callback(err);
+                    callback(null, results);
+                    });
+                }else {
+                    console.log(`The id ${id} you added is not found in database! enter a valid id!`);
+                }
+            })
+            .catch(err => console.log(err));
     }
 
     delete(id, callback) {
         // Write code and query to delete an existing TODO item
+        const deletePromise = this.availableIds(this.sqlSelectQuery);
+        deletePromise
+            .then(data => {
+                const temp = data.includes(id);
+                if(temp) {
+                    const deleteItem = `DELETE FROM todo_items WHERE id =` + dbConnection.escape(id);
+                    this.dbConnection.query(deleteItem, (err, results, fields) => {
+                    if(err) return callback(err);
+                    callback(null, results);
+        });
+                }else {
+                    console.log(`The id ${id} you added is not found in database! enter a valid id!`);
+                }
+            })
+            .catch(err => console.log(err));
     }
 
     tagTodoItem(todoItemId, tagId, callback) {
         // Write code and query add a tag to a TODO item
+        const allItemsId = this.availableIds(this.sqlSelectQuery);
+        allItemsId
+            .then(data => {
+                const todoIdIsFound = data.includes(todoItemId);
+                if(todoIdIsFound) {
+                    const checkTagId = this.availableIds("SELECT * FROM tags");
+                    checkTagId
+                        .then(tagData => {
+                            const tagIdIsFound = tagData.includes(tagId);
+                            if(tagIdIsFound) {
+                                const addTodoItemTag = "INSERT INTO todo_item_tag VALUES(?, ?)";
```

Comment:
the previous checks are a little redundant and NOT needed on sql, remember this is not a procedural language, its a declarative one. You query all the data and checking in memory if an id exists, the database can do all of this operations way faster and without using the memory. 

you could have checked this in sql with something like 
```select count(1)
from table
where key = value;```

but its not even needed because of the foreign key constraints, if the foreign keys don't exist the record won't be inserted